lib/utils/compose.coffee

```diff
@@ -0,0 +1,732 @@
+			boolean: true
+			alias: 'e'
+		},
+		{
+			signature: 'inlinelogs'
+			description: 'Display all log output from builds inline as a stream'
+			boolean: true
+		},
+	]
+
+exports.generateOpts = (options) ->
+	fs = require('mz/fs')
+	fs.realpath(options.source || '.').then (projectPath) ->
+		projectName: options.projectName
+		projectPath: projectPath
+		inlineLogs: !!options.inlinelogs
+
+defaultProjectTemplate = (serviceName, image) ->
+	image = "image: #{image}" if image
+	"""
+	version: '2.1'
+	networks: {}
+	volumes:
+	  resin-app-#{serviceName}: {}
+	services:
+	  #{serviceName}:
+	    build: '.'
+	    #{image}
+	    privileged: true
+	    restart: unless-stopped
+	    network_mode: host
+	    volumes:
+	      - resin-app-#{serviceName}:/data
+	    labels:
+	      io.resin.features.kernel_modules: 1
+	      io.resin.features.firmware: 1
+	      io.resin.features.dbus: 1
+	      io.resin.features.supervisor_api: 1
+	      io.resin.features.resin_api: 1
+	      io.resin.update.strategy: download-then-kill
+	      io.resin.update.handover_timeout: ""
+	"""
+
+compositionFileNames = [
+	'resin-compose.yml'
+	'resin-compose.yaml'
+	'docker-compose.yml'
+	'docker-compose.yaml'
+]
+
+# look into the given directory for valid compose files and return
+# the contents of the first one found.
+resolveProject = (rootDir) ->
+	fs = require('mz/fs')
+	Promise.any compositionFileNames.map (filename) ->
+		fs.readFile(path.join(rootDir, filename), 'utf-8')
+
+# Parse the given composition and return a structure with info. Input is:
+#  - composePath: the *absolute* path to the directory containing the compose file
+#  - composeStr: the contents of the compose file, as a string
+createProject = (composePath, composeStr, projectName = null) ->
+	yml = require('js-yaml')
+	compose = require('resin-compose-parse')
+
+	Promise.try ->
+		composition = yml.safeLoad(composeStr, schema: yml.FAILSAFE_SCHEMA)
+		compose.normalize(composition)
+	.then (composition) ->
+		projectName ?= path.basename(composePath)
+		descriptors = compose.parse(composition).map (descr) ->
+			# generate an image name based on the project and service names
+			# if one is not given and the service requires a build
+			if descr.image.context? and not descr.image.tag?
+				descr.image.tag = [ projectName, descr.serviceName ].join('_')
+			return descr
+		return {
+			path: composePath,
+			name: projectName,
+			composition,
+			descriptors
+		}
+
+# high-level function resolving a project and creating a composition out
+# of it in one go. if image is given, it'll create a default project for
+# that without looking for a project. falls back to creating a default
+# project if none is found at the given projectPath.
+exports.loadProject = (logger, projectPath, projectName, appName, image) ->
+	logger.logDebug('Loading project...')
+
+	Promise.try ->
+		if image?
+			logger.logInfo("Creating default resin-compose project with image: #{image}")
```

Comment:
Yeah, I reused docker-compose terminology here because we lacked our own. Project is merely the folder which contains, well, a docker-compose.yml and related folders with Dockerfiles, etc. Happy to change to whatever we like (at least the most prominent use which is the option named `--project`). cc @shaunmulligan